File: typescript/LWTSD/ResourceTypes/ResourceBase64Binary.ts
var xml = require('../../node-xmpp/packages/xml');

import { LWTSD } from '../index';
import { SimplifiedType } from '../SimplifiedType';
import { ResourceDescription } from '../ResourceDescription'
import { ResourceRestriction } from '../ResourceRestriction';
import{ Resource } from './Resource'

export class ResourceBase64Binary extends Resource
{
    public readonly HistoricalValues :  Array< [Date, string] > = new Array< [Date, string] >();

    InnerValue : string;

    public GetValue() : string
    {
        return this.InnerValue;
    }

    public SetValue(Value : string) : void
    {
        this.InnerValue = Value;
        this.ModifiedAt = new Date();
    }

    public constructor(Element? : any)
    {
        super(Element);

        let points = Element.getChildren("point");
        for (let i = 0; i < points.length; i++)
        {
            let it = points[i];
            let TimeStamp = new Date(it.attrs.timestamp);
            let PointValue = it.text();
            this.HistoricalValues.push( [TimeStamp, PointValue] );
            if (TimeStamp > this.ModifiedAt)
            {
                this.ModifiedAt = TimeStamp;
                this.InnerValue = PointValue;
            }
        }
    }

    public GetPoint() : any
    {
        let RVal = xml("point", { /*"xmlns": LWTSD.Namespace */}, this.GetValue());
        RVal.attrs.timestamp = this.ModifiedAt;
        return RVal;
    }

    public GetWriteValue() : string
    {
        return this.GetValue();
    }

    public GetDescription() : ResourceDescription
    {
        return super.GetDescriptionFromSimpleType(SimplifiedType.Base64Binary);
    }

    public LoadFromWrite(WriteElement : any) : void
    {
        this.SetValue( WriteElement.text() );
    }
}

File: NET/LWTSD/ResourceTypes/ResourceBase64Binary.cs
using System;
using System.Xml;
using System.Xml.Linq;
using System.Collections.Generic;

namespace LWTSD.ResourceTypes
{
	public class ResourceBase64Binary : Resource
	{
		public readonly List<Tuple<DateTime, string>> HistoricalValues = null;

		string InnerValue;

		public string Value
		{
			get
			{
				return InnerValue;
			}

			set
			{
				InnerValue = value;
				ModifiedAt = DateTime.UtcNow;
			}
		}

		public ResourceBase64Binary()
		{
		}

		public ResourceBase64Binary(XElement Element)
			: base(Element)
		{
			foreach (var it in Element.Elements(LWTSD.Namespace + "point"))
			{
                if (HistoricalValues == null)
                    HistoricalValues = new List<Tuple<DateTime, string>>();

                DateTime TimeStamp = XmlConvert.ToDateTime(it.Attribute("timestamp").Value,
														   XmlDateTimeSerializationMode.Utc);
				string PointValue = it.Value;
				HistoricalValues.Add(new Tuple<DateTime, string>(TimeStamp, PointValue));
				if (TimeStamp > ModifiedAt)
				{
					ModifiedAt = TimeStamp;
					InnerValue = PointValue;
				}
			}
		}

		public override XElement GetPoint()
		{
			XElement Point = base.GetPoint();
			Point.Value = Value;
			return Point;
		}

		public override string GetWriteValue()
		{
			return Value;
		}

		public override ResourceDescription GetDescription()
		{
			return base.GetDescription(SimplifiedType.Base64Binary);
		}

		public override void LoadFromWrite(XElement WriteElement)
		{
			this.Value = WriteElement.Value;
		}
	}
}
